compiler/rustc_middle/src/ty/fast_reject.rs

```diff
@@ -182,3 +184,188 @@ impl<D: Copy + Debug + Eq> SimplifiedTypeGen<D> {
+        | ty::Ref(..)
+        | ty::Never
+        | ty::Tuple(..)
+        | ty::FnPtr(..)
+        | ty::Foreign(..)
+        | ty::Opaque(..) => {}
+        ty::FnDef(..)
+        | ty::Closure(..)
+        | ty::Generator(..)
+        | ty::GeneratorWitness(..)
+        | ty::Placeholder(..)
+        | ty::Bound(..)
+        | ty::Infer(_) => bug!("unexpected impl_ty: {impl_ty}"),
+    }
+
+    let k = impl_ty.kind();
+    match *obligation_ty.kind() {
+        // Purely rigid types, use structural equivalence.
+        ty::Bool
+        | ty::Char
+        | ty::Int(_)
+        | ty::Uint(_)
+        | ty::Float(_)
+        | ty::Str
+        | ty::Never
+        | ty::Foreign(_) => obligation_ty == impl_ty,
+        ty::Ref(_, obl_ty, obl_mutbl) => match k {
+            &ty::Ref(_, impl_ty, impl_mutbl) => {
+                obl_mutbl == impl_mutbl && types_may_unify(obl_ty, impl_ty)
+            }
+            _ => false,
+        },
+        ty::Adt(obl_def, obl_substs) => match k {
+            &ty::Adt(impl_def, impl_substs) => {
+                obl_def == impl_def
+                    && iter::zip(obl_substs, impl_substs)
+                        .all(|(obl, imp)| generic_args_may_unify(obl, imp))
+            }
+            _ => false,
+        },
+        ty::Slice(obl_ty) => matches!(k, &ty::Slice(impl_ty) if types_may_unify(obl_ty, impl_ty)),
+        ty::Array(obl_ty, obl_len) => match k {
+            &ty::Array(impl_ty, impl_len) => {
+                types_may_unify(obl_ty, impl_ty) && consts_may_unify(obl_len, impl_len)
+            }
+            _ => false,
+        },
+        ty::Tuple(obl) => match k {
+            &ty::Tuple(imp) => {
+                obl.len() == imp.len()
+                    && iter::zip(obl, imp).all(|(obl, imp)| types_may_unify(obl, imp))
+            }
+            _ => false,
+        },
+        ty::RawPtr(obl) => match k {
+            ty::RawPtr(imp) => obl.mutbl == imp.mutbl && types_may_unify(obl.ty, imp.ty),
+            _ => false,
+        },
+        ty::Dynamic(obl_preds, ..) => {
+            // Ideally we would walk the existential predicates here or at least
+            // compare their length. But considering that the relevant `Relate` impl
+            // actually sorts and deduplicates these, that doesn't work.
+            matches!(k, ty::Dynamic(impl_preds, ..) if
+                obl_preds.principal_def_id() == impl_preds.principal_def_id()
+            )
+        }
+        ty::FnPtr(obl_sig) => match k {
+            ty::FnPtr(impl_sig) => {
+                obl_sig.abi() == impl_sig.abi()
+                    && obl_sig.unsafety() == impl_sig.unsafety()
+                    && obl_sig.inputs_and_output().skip_binder().len()
+                        == impl_sig.inputs_and_output().skip_binder().len()
+                    && iter::zip(
+                        obl_sig.inputs_and_output().skip_binder(),
+                        impl_sig.inputs_and_output().skip_binder(),
```

Comment:
Lots of repetition here that can be factored out.

Also, would it be better to pattern match the contents of the `FnPtr`? That way if an additional field is ever added it won't be missed here.